src/core/Akka/Event/LogFilter.cs

```diff
@@ -0,0 +1,314 @@
+ */
+
+public enum LogFilterType
+{
+    /// <summary>
+    /// Filter log messages based on their source
+    /// </summary>
+    Source,
+
+    /// <summary>
+    /// Filter log messages based on their message
+    /// </summary>
+    Message
+}
+
+public enum LogFilterDecision
+{
+    Keep,
+    Drop,
+
+    /// <summary>
+    /// If we're asked to evaluate a filter and we don't have enough information to make a decision.
+    ///
+    /// For instance: a <see cref="LogFilterType.Source"/> stage gets asked to evaluate a message body.
+    /// </summary>
+    NoDecision
+}
+
+/// <summary>
+/// Base class for all log filters
+/// </summary>
+/// <remarks>
+/// Worth noting: these run inside the Logging actors, so they're out of band
+/// from any high performance workloads already.
+///
+/// In addition to this - all log filters will only run if the log level is enabled.
+///
+/// i.e. if we're at INFO level and the filter is set to a lower level, i.e. filtering DEBUG
+/// logs, the filter won't even run.
+/// </remarks>
+public abstract class LogFilterBase : INoSerializationVerificationNeeded, IDeadLetterSuppression
+{
+    /// <summary>
+    /// Which part of the log message this filter is evaluating?
+    /// </summary>
+    /// <remarks>
+    /// This actually has a performance implication - if we're filtering on the source, which
+    /// is already fully "expanded" into its final string representation, we can try to fail fast
+    /// on that without any additional allocations.
+    ///
+    /// If we're filtering on the message, we have to fully expand the log message first which
+    /// involves allocations. Users on really tight performance budgets should be aware of this.
+    /// </remarks>
+    public abstract LogFilterType FilterType { get; }
+
+    /// <summary>
+    /// Fast path designed to avoid allocating strings if we're filtering on the message content.
+    /// </summary>
+    /// <param name="part">The part of the message to evaluate.</param>
+    /// <param name="content">Usually the fully expanded message content.</param>
+    public abstract LogFilterDecision ShouldKeepMessage(LogFilterType part, string content);
+}
+
+/// <summary>
+/// Uses a regular expression to filter log messages based on their source.
+/// </summary>
+public sealed class RegexLogSourceFilter : LogFilterBase
+{
+    private readonly Regex _sourceRegex;
+
+    public RegexLogSourceFilter(Regex sourceRegex)
+    {
+        _sourceRegex = sourceRegex;
+    }
+
+    public override LogFilterType FilterType => LogFilterType.Source;
+
+    public override LogFilterDecision ShouldKeepMessage(LogFilterType part, string content)
+    {
+        if (part == LogFilterType.Source)
+            return _sourceRegex.IsMatch(content) ? LogFilterDecision.Drop : LogFilterDecision.Keep;
+        return LogFilterDecision.NoDecision;
+    }
+}
+
+public sealed class ExactMatchLogSourceFilter : LogFilterBase
```

Comment:
I might get rid of this - because the name of the `ActorSystem` _usually_ gets appended to the `LogSource`, therefore I think this will never work properly for most users. But, you never know.